[Haleema Irfan] iP

haleemairfan opened the pull request:
### Meet Derek!

> Derek saved my dog from getting run over - Dr Doofenshmirtz

Derek, your best friend, your buddy, your saviour is here to brighten up your life!

With Derek you can
- Schedule your very real social life
- Sort your life out with a to-do list!
- Finally have a friend

Wow! Sounds like a dream doesn't it 🫣
Well this is real, Derek even celebrates your little achievements! 
```java
        String[] celebrationMessages = new String[]{"yay!",
            "woohoo!",
            "let's go!!!!",
            "great job :)",
            "you're on a roll!"};
```
What a gentleman, better than a real man for sure!
You can have him with a few easy steps!

1. Download him [here](https://github.com/haleemairfan/ip/releases/tag/A-Jar) 
    _Don't tell him you downloaded him, he thinks he's real_
2. Double click it!!!!!!
3. Start yapping with him

Derek's quirks

- [ ] managing your tasks
- [ ] managing your events
- [x] managing your deadlines
- [ ] just a great guy in general

**So what are you waiting for!!! Get Derek NOWWWW**
_Please get him I need to get rid of him._

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

haleema-irfan committed:
level 0

haleema-irfan committed:
level 1

haleema-irfan committed:
level3

haleema-irfan committed:
level 4 and 5

haleema-irfan committed:
level 6 and level 3

haleema-irfan committed:
A-TextUiTesting v1

haleema-irfan committed:
Add javadoc comments

haleema-irfan committed:
Add instructions for user

haleema-irfan committed:
Add testing

haleema-irfan committed:
Add text-based testing

haleema-irfan committed:
Add Junit testing

haleema-irfan committed:
Remove buggy Junit tests

haleema-irfan committed:
Level 2

haleema-irfan committed:
A-TextUiTesting

haleema-irfan committed:
Level-5

haleema-irfan committed:
Level-6

haleema-irfan committed:
Level-6

haleema-irfan committed:
Add Level-0

haleema-irfan committed:
Add Level-0

haleema-irfan committed:
Add Level-0

haleema-irfan committed:
Add Level-1

haleema-irfan committed:
Add Level-2

haleema-irfan committed:
Add Levels 2 to 6, along with testing

haleema-irfan committed:
Level 7: Save tasks to disk

Current situation: Tasks managed by the chatbot are not persisted between sessions, leading to a loss of data whenever the application is restarted.

Why it needs to change: Users need their task data to be saved and loaded automatically to ensure continuity and avoid manual re-entry of tasks each time the chatbot is used.

What is being done about it: Implement automatic saving of the task list to the hard disk whenever any changes are made. Load the task data from the disk when the chatbot starts up.

Why it is done that way: Hard-coding the file path (./data/derek.txt) and saving the tasks in a simple text format ensures a straightforward and reliable method of persistence that aligns with the current project scope.

Any other relevant info: The file format includes task type, completion status, task details, and any relevant timestamps. This format was chosen for its simplicity and ease of parsing.

haleema-irfan committed:
Merge Level 7 branch into master

haleema-irfan committed:
Level 8: Implement date and time handling for tasks

Current situation: The chatbot currently treats date and time inputs as plain strings without any understanding of their actual meaning or structure.

Why it needs to change: Users need the chatbot to recognize and manage deadlines and events with accurate date and time information, allowing for more meaningful task management and scheduling.

What is being done about it: Implement the ability to parse and store dates and times as LocalDateTime objects in task-related classes. Accept input in formats like yyyy-MM-dd and output in a user-friendly format such as MMM dd yyyy hh:mm a. Additionally, add support for recognizing and formatting date and time inputs correctly when creating tasks.

Why it is done that way: Using LocalDateTime allows for robust and flexible handling of date and time data, ensuring tasks can be managed accurately. It also enables future extensions, such as filtering tasks by date or checking for upcoming deadlines.

Any other relevant info: This change lays the groundwork for further enhancements, such as commands to filter or list tasks based on specific dates, improving the overall functionality and user experience of the chatbot.

haleema-irfan committed:
Merge branch-Level-8 into master

haleema-irfan committed:
Add more OOP

haleema-irfan committed:
Refactor Code to Extract Core Functionalities into Separate Classes

Current Situation:
The current code structure has multiple responsibilities handled within a single class, leading to a tightly coupled and less maintainable codebase. The code lacks clear separation of concerns, making it harder to understand, test, and extend.

Why It Needs to Change:
The lack of modularity increases the complexity of the code, making it difficult to manage and maintain. By having multiple responsibilities in one place, the code violates the Single Responsibility Principle, which can lead to more bugs and challenges in implementing future changes or enhancements.

What Is Being Done About It:
Refactor the code to extract out closely related functionalities into dedicated classes:
- Ui: Handles all interactions with the user.
- Storage: Manages the loading and saving of tasks from and to a file.
- Parser: Interprets and processes user commands.
- TaskList: Manages the task list, including operations to add or delete tasks.

Why It Is Done That Way:
This refactoring adheres to the principles of Object-Oriented Programming (OOP), particularly the Single Responsibility Principle. By extracting specific responsibilities into their respective classes, the code becomes more modular, easier to understand, and more straightforward to test. This approach also lays a solid foundation for future enhancements and ensures that each class is focused on a single responsibility.

haleema-irfan committed:
Organize classes into suitable packages

haleema-irfan committed:
Add-gradle-support

haleema-irfan committed:
Add-gradle-support

haleema-irfan committed:
Add J-Unit tests

haleema-irfan committed:
Add error checking for list size in commands

haleema-irfan committed:
Add Java Doc Comments

haleema-irfan committed:
Implement Coding Standard

haleema-irfan committed:
Add Level-9

haleema-irfan committed:
Resolve merge conflicts

haleema-irfan committed:
Resolve merge conflicts

haleema-irfan committed:
Add checkstyle

isaactodo commented on src/main/java/derek/Parser.java:
```diff
@@ -0,0 +1,138 @@
+    private Ui ui;
+
+    /**
+     * Constructs a {@code Parser} object with the specified command, storage, and UI.
```
I like how you used @code tags in the JavaDoc

isaactodo commented on src/main/java/derek/Parser.java:
```diff
@@ -0,0 +1,138 @@
+                    throw new IncorrectCommandException("Please enter your commands correctly for Derek "
+                            + "(e.g. todo (task)), he keeps throwing tantrums");
+                }
+                this.getTask();
```
Any reason why you didn't use switch statements here? I feel the code would look cleaner if you used switches.

haleemairfan replied:
Hi! thanks for the suggestion, the use of switch statements slipped my mind. I will definitely be adding them :)

isaactodo commented on src/main/java/derek/Parser.java:
```diff
@@ -0,0 +1,138 @@
+                throw new IncorrectCommandException("Is it a todo, event, or deadline?\n"
+                        + "Please enter your commands correctly for Derek (e.g. todo (task)), "
+                        + "he keeps throwing tantrums");
+            }
```
Can try using switch statements.

isaactodo commented on src/main/java/derek/Parser.java:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * Constructs a {@code Parser} object with the specified command, storage, and UI.
+     *
+     * @param command the user command input
+     * @param storage the storage object containing the task list
+     * @param ui the UI object to interact with the user
```
Missing punctuation behind each parameter description.

isaactodo commented on src/main/java/derek/Storage.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Writes a given string to the task data file.
+     *
+     * @param textToAdd the string to be added to the file
+     * @throws IOException if an I/O error occurs during file writing
```
Missing punctuation behind @param description and @throws description. I noticed the same issue in several other places too.
```suggestion
     * @param textToAdd the string to be added to the file.
     * @throws IOException if an I/O error occurs during file writing.
```

isaactodo commented on src/main/java/derek/Ui.java:
```diff
@@ -0,0 +1,172 @@
+        this.isRunning = false;
+    }
+
+    public void returnList() {
```
I feel that printList() would be a more suitable name for this method.

isaactodo commented on src/main/java/derek/command/Command.java:
```diff
@@ -0,0 +1,44 @@
+ * based on the input string. It also provides methods to check the validity of the command
+ * and to extract the task details from the command.
+ */
+public class Command {
```
I feel that separating out each command into its own separate class generates a lot of boilerplate code, and you have the added hassle of having to instantiate each class. An alternative is to use Enums.

haleemairfan replied:
Sorry, how would using Enums work?

isaactodo commented on src/main/java/derek/command/CompleteCommand.java:
```diff
@@ -0,0 +1,36 @@
+ * It extends the {@code Command} class and executes the command to complete a task.
+ */
+public class CompleteCommand extends Command {
+
```
Why is there a blank line here? Similar issues in the other command files.

haleemairfan replied:
Thanks for pointing these issues out! I will definitely go through and fix any issues with the coding standard.

isaactodo commented on src/test/java/derek/ParserTest.java:
```diff
@@ -0,0 +1,41 @@
+import derek.task.TaskList;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Incorrect import ordering.

isaactodo commented on src/test/java/derek/UiTest.java:
```diff
@@ -0,0 +1,61 @@
+import java.io.InputStream;
+import java.io.PrintStream;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
Incorrect import ordering.

isaactodo commented on src/main/java/derek/Ui.java:
```diff
@@ -0,0 +1,172 @@
+        System.out.println("Here you are!" + "\n" + task);
+    }
+
+    public String generateRandomCelebration() {
```
Interesting feature! 🥳

isaactodo reviewed (commented):
LGTM, just a few nits to fix 👍

haleema-irfan committed:
Add JavaFX functionality

haleema-irfan committed:
Add JavaFX functionality

haleema-irfan committed:
Debug JavaFX GUI

cweijin commented on src/main/java/derek/Storage.java:
```diff
@@ -0,0 +1,145 @@
+
+    }
+
+    /**
+     * Overrides the task data file by clearing its contents.
+     *
+     * @throws IOException if an I/O error occurs during file writing
+     */
+    public void overrideFile() throws IOException {
+        FileWriter writer = new FileWriter(dataFilePathName);
+        writer.write("");
+        writer.close();
+
+    }
```
Perhaps `overwriteFile()`?

haleemairfan replied:
Yes, that is much clearer! Thanks for the suggestion :)

cweijin commented on src/main/java/derek/Storage.java:
```diff
@@ -0,0 +1,145 @@
+                throw new FileNotFoundException("Data file not found");
+            } else if (tasks.isDirectory()) {
+                throw new FileNotFoundException("Path is a directory");
+            }
```
I like how you handled different type of situation that might occurs when reading a filePath

cweijin commented on src/main/java/derek/Ui.java:
```diff
@@ -0,0 +1,187 @@
+        this.user = name.nextLine();
+        initiateUserInteraction();
+         */
+    }
```
This is cute! 😻

cweijin commented on src/main/java/derek/Ui.java:
```diff
@@ -0,0 +1,187 @@
+        int randomNumber = random.nextInt((max - min) + 1) + min;
+        String celebration = celebrationMessages[randomNumber];
+        return celebration;
+    }
```
So much fun!! 😆

cweijin commented on src/main/java/derek/Ui.java:
```diff
@@ -0,0 +1,187 @@
+
+
+
+
```
Perhaps removing these trailing spaces

cweijin reviewed (commented):
LGTM! 👍 
Good naming convention, sensible, not too long / short, making code very readable.

haleema-irfan committed:
Add A-Enums

haleema-irfan committed:
Add assertions to ensure valid task list and file operations in Derek bot

To improve code reliability and catch potential issues early, assertions have
been added to various parts of the Derek bot codebase. These assertions
document critical assumptions at runtime, allowing for better debugging during
development.

- Added assertions in Derek class to ensure task list is not null after being
  loaded from storage.
- Added checks to ensure file operations (e.g., reading and overriding the file)
  succeed, using assertions to confirm correct behavior.
- Parser class now includes assertions that ensure task list size is valid and
  non-negative.

These changes help identify potential errors early while maintaining
correct task list operations and storage interactions.

haleema-irfan committed:
Add assertions to ensure valid task list and file operations in Derek bot

The current codebase lacks assertions to validate key assumptions, which could lead to unhandled errors during runtime, particularly in task list management and file operations.

This needs to change to improve code reliability and ensure potential issues are caught early during development.

Add assertions to critical areas of the codebase, such as verifying the task list is not null after loading from storage and ensuring file operations succeed. Include assertions in the Parser class to validate that the task list size is non-negative.

This is done to document and enforce assumptions at runtime, improving debugging and making it easier to identify potential problems.

By adding these checks, we ensure that task list operations and storage interactions behave as expected, reducing the likelihood of unnoticed errors.

haleema-irfan committed:
Refactor codebase to follow SLAP and improve structure

The codebase contains several long and deeply nested methods, making the code difficult to read, maintain, and debug.

This needs to change to improve modularity, maintainability, and readability of the code.

Refactor the affected methods by splitting them into smaller, focused methods. Reduce complexity and eliminate arrowhead-style nested code.

This is done to adhere to the Single Level of Abstraction Principle (SLAP) and ensure better separation of concerns, allowing for easier maintenance and updates.

All refactored classes follow best practices, resulting in a cleaner and more maintainable codebase.

haleema-irfan committed:
**Refactor codebase to follow SLAP and improve structure**

Refactored multiple classes across the codebase to improve readability,
maintainability, and code quality. The changes adhere to the Single
Level of Abstraction Principle (SLAP), reduce the complexity of methods,
and eliminate arrowhead-style nested code.

Previously, several methods were too long and nested, making the code
difficult to read and maintain. By splitting these into smaller, more
focused methods, we improved modularity and clarity across the code.

Key changes include breaking down command processing logic, reducing
nested switch-case statements, and ensuring that methods follow a clean,
concise flow. This refactor enhances the code's structure and makes it
easier to extend and maintain in the future.

haleemairfan committed:
Merge pull request #1 from haleemairfan/branch-A-Assertions

Add assertions to ensure valid task list and file operations in Derek…

haleema-irfan committed:
Resolved merge conflicts caused by merging master into A-CodeQuality branch.

Merged the latest master branch into A-CodeQuality, resolving conflicts in:
- src/main/java/derek/Ui.java
- src/main/java/derek/Command

Retained master updates while preserving code quality improvements from A-CodeQuality. Aligned method signatures and resolved logic conflicts. Tested to ensure functionality after the merge.

haleema-irfan committed:
Add functionality to display task completion statistics for the past week

Currently, the app does not provide a way to view statistics on tasks, such as how many have been completed in the past week.

Tracking task completion over time is crucial for users to monitor their productivity and gain insights into their progress.

Implement a feature that calculates and displays the number of tasks completed in the last 7 days.

Displaying weekly task completion statistics helps users track short-term progress and fosters better time management.

The feature integrates with the existing task list and formats the output to ensure clarity and ease of understanding.

haleema-irfan committed:
Resolve merge conflicts

haleema-irfan committed:
Improve code quality

haleema-irfan committed:
Add instructions

haleema-irfan committed:
Add JavaDoc comments

haleema-irfan committed:
Improve naming conventions

haleema-irfan committed:
Organise packages

haleema-irfan committed:
Improve error handling for task input

haleema-irfan committed:
Resolve merge conflicts from the merging of branch-A-Assertions to master branch

haleema-irfan committed:
Resolve merge conflicts from merging branch-A-CodeQuality to master

haleemairfan committed:
Add User Guide

haleemairfan committed:
Add files via upload

haleema-irfan committed:
Add Better GUI

haleema-irfan committed:
Add resizable GUI

haleema-irfan committed:
Add Main Window CSS

haleema-irfan committed:
Add Dialog Box CSS

haleema-irfan committed:
Add custom dialog box

haleema-irfan committed:
Add images

haleema-irfan committed:
Improve GUI

haleemairfan committed:
Update README.md

haleemairfan committed:
Update README.md

haleemairfan committed:
Delete docs/Ui.png

haleemairfan committed:
Add UI.png

haleema-irfan committed:
Improve button

haleemairfan committed:
Update README.md

haleemairfan committed:
Update README.md

haleemairfan committed:
Update README.md

haleema-irfan committed:
Fix delete UI

haleema-irfan committed:
Fix UI

haleemairfan committed:
Delete data directory

haleema-irfan committed:
Add deleted JUnit tests

haleema-irfan committed:
Add JUnit Tests
